types/src/block.rs

```diff
@@ -0,0 +1,324 @@
+use tendermint::block::parts::Header as PartSetHeader;
+use tendermint::block::{Commit, CommitSig, Header, Id};
+use tendermint::signature::SIGNATURE_LENGTH;
+use tendermint::vote;
+use tendermint::{chain, Hash, Vote};
+
+use crate::consts::{genesis::MAX_CHAIN_ID_LEN, version};
+use crate::{Error, Result, ValidateBasic, ValidationError, ValidationResult};
+
+trait IsZero {
+    fn is_zero(&self) -> bool;
+}
+
+impl IsZero for Id {
+    fn is_zero(&self) -> bool {
+        matches!(self.hash, Hash::None) && self.part_set_header.is_zero()
+    }
+}
+
+impl IsZero for PartSetHeader {
+    fn is_zero(&self) -> bool {
+        matches!(self.hash, Hash::None) && self.total == 0
+    }
+}
+
+impl ValidateBasic for Header {
+    fn validate_basic(&self) -> ValidationResult<()> {
+        if self.version.block != version::BLOCK_PROTOCOL {
+            return Err(ValidationError::IncorrectBlockProtocol(self.version.block));
+        }
+
+        if self.chain_id.as_str().len() > MAX_CHAIN_ID_LEN {
+            return Err(ValidationError::ChainIdTooLong(
+                self.chain_id.as_str().into(),
+            ));
+        }
+
+        if self.height.value() == 0 {
+            return Err(ValidationError::ZeroHeight);
+        }
+
+        if self.last_block_id.is_none() && self.height.value() != 1 {
+            return Err(ValidationError::MissingLastBlockId);
+        }
```

Comment:
Good catch, the go implementation is [here](https://github.com/celestiaorg/celestia-core/blob/v0.34.x-celestia/types/block.go#L64) and it just allows for this field to be a `Zero`. In case of it being `Zero`, `tendermint-rs` will deserialize it as a `None`. So the go implementation is permissive there.
I empirically checked that this is `Zero` only in case where the block is a genesis and went with this stricter variant of not allowing it to be missing in other cases, but we can remove it to just mimic the Go behavior 1:1

Reply:
as we discussed, introduced a constant for the height of the genesis block